clients/drcachesim/tools/histogram.cpp

```diff
@@ -135,27 +155,39 @@ cmp(const std::pair<addr_t, uint64_t> &l, const std::pair<addr_t, uint64_t> &r)
 }
 
 bool
-histogram_t::print_results()
+histogram_t::reduce_results(uint64_t *unique_icache_lines, uint64_t *unique_dcache_lines)
 {
-    shard_data_t total;
     if (shard_map_.empty()) {
-        total = serial_shard_;
+        reduced_ = serial_shard_;
     } else {
         for (const auto &shard : shard_map_) {
             for (const auto &keyvals : shard.second->icache_map) {
-                total.icache_map[keyvals.first] += keyvals.second;
+                reduced_.icache_map[keyvals.first] += keyvals.second;
             }
             for (const auto &keyvals : shard.second->dcache_map) {
-                total.dcache_map[keyvals.first] += keyvals.second;
+                reduced_.dcache_map[keyvals.first] += keyvals.second;
             }
         }
     }
+    if (unique_icache_lines != nullptr)
+        *unique_icache_lines = reduced_.icache_map.size();
+    if (unique_dcache_lines != nullptr)
+        *unique_dcache_lines = reduced_.dcache_map.size();
+    return true;
+}
+
+bool
+histogram_t::print_results()
+{
+    if (!reduce_results())
```

Comment:
`reduce_results()` always returns `true` which makes this check irrelevant unless we update `reduce_results()` with a path that returns `false`.

Reply:
Yes, it is forward-thinking, and it is virtual so a child class can override it and extend it with something that could fail.